### PR TITLE
Clean up "default" service port handling.

### PR DESCRIPTION
**Pull request by brendandburns**
See #983 for context.  Builds on #4800 (though that's not required, if we decide we don't like the name)

### CODE REVIEW

**Comment by brendandburns** [comment]
Also see #3838, for the motivation for using the service port, if the ports list on the pod is empty.

**Comment by vishh** [comment]
LGTM. @brendandburns did you run e2e tests?

**Comment by brendandburns** [comment]
Ok, removed the field rename, since @thockin says he's doing it as part of his refactor.

**Comment by thockin** [comment]
Will review this today

**Comment by brendandburns** [comment]
Yes, ran e2e, everything passed except cAdvisor, since I was sync'd to a time when that test was broken, I believe that fail is unrelated to this PR.  I can rebase and re-run if you would like me to.

**Comment by brendandburns** [comment]
@thockin ping for review ;)

File: pkg/service/endpoints_controller.go [hunk 1]
```diff
@@ -150,18 +150,35 @@ func endpointsEqual(eps *api.Endpoints, endpoints []api.Endpoint) bool {
...
-	firstContainerPort := 0
-	if len(pod.Spec.Containers) > 0 && len(pod.Spec.Containers[0].Ports) > 0 {
-		firstContainerPort = pod.Spec.Containers[0].Ports[0].ContainerPort
+func findDefaultPort(pod *api.Pod, servicePort int) (int, bool) {
+	foundPort := 0
+	portMap := map[string]int{}
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			if port.Name == "" {
+				foundPort = port.ContainerPort
```
Comment (thockin) [inline pkg/service/endpoints_controller.go]: Wouldn't you want to break here, but you probably want a double break (break-to-label) ?

File: pkg/service/endpoints_controller.go [hunk 2]
```diff
@@ -150,18 +150,35 @@ func endpointsEqual(eps *api.Endpoints, endpoints []api.Endpoint) bool {
...
+			} else {
+				portMap[port.Name] = port.ContainerPort
+			}
+		}
 	}
+	if foundPort == 0 && len(portMap) == 0 {
+		return servicePort, true
+	}
+	if port, found := portMap["default"]; found {
```
Comment (thockin) [inline pkg/service/endpoints_controller.go]: If the only name we ever look up is "default" why build the map at all?

**Comment by thockin** [comment]
I do not understand this, I find the logic a bit twisty.

What is the meaning of the new "default" magic?

This is a semantic change and not strictly compatible.  If we're going to break compat, let's just go all the way and say that if you don't specify a ContainerPort you get the ServicePort, full stop.  No "default" magic..

**Comment by brendandburns** [comment]
The thing about "default" is that you actually won't break compat for 99.9% of use cases, since they only have a single port anyway.

The logic is actually pretty straightforward:

If you name a port, you get that port, or an error if it doesn't exist.

If you don't name a port:
  If the pod has no ports, you get the service port on that pod
  If the pod has one port, you get that port
  If the pod has multiple ports, you get the one with no name, or the name of "default"

I think that this is actually a good compromise that will provide compat to 99% of existing users, while moving to something more sane and explicable.

**Comment by thockin** [comment]
On Tue, Mar 3, 2015 at 2:19 PM, Brendan Burns <notifications@github.com> wrote:
>
> The thing about "default" is that you actually won't break compat for 99.9% of use cases, since they only have a single port anyway.
>
> The logic is actually pretty straightforward:
>
> If you name a port, you get that port, or an error if it doesn't exist.
>
> If you don't name a port:
> If the pod has no ports, you get the service port on that pod

This covers a corner case that was not covered before, so I'm in favor.

> If the pod has one port, you get that port

This is current behavior, agreed.

> If the pod has multiple ports, you get the one with no name, or the name of "default"

It's the last part that I don't get.  Why?  Why is ["default"] better
than [0] ?  It adds a special case and code to support it to solve a
problem that I don't think exists - is there somewhere that people
complain about [0] being bad?

Especially if I get my own wish and we just simplify to "If you don't
name a destination port, you get the service port on that pod" which
makes all of the "default" stuff go away.

**Comment by brendandburns** [comment]
We have a bug open to make the port handling cleaner, I'd prefer to merge
this pr, close that bug and the rip it out again if your code lands, than
wait.  But I don't feel super strongly about it.  I would def. like to fix
the no ports on the pod case.

Brendan

**Comment by thockin** [comment]
Could we keep the no-ports case here and just drop the "default" stuff?

**Comment by brendandburns** [comment]
I ripped out the default port stuff.  I will reming @thockin that he suggested that port[0] was confusing in the first place.

Given the feedback on this PR, I'm going to close #983 as WAI, if we want to kill named ports we can do that with a different issue.

**Comment by thockin** [comment]
For the sake of progress I'm going to roll with this, though I am not sure why you classify the case of default port (not specified) and multiple ports on a pod as an error?

Yes the [0] semantic is confusing and awful, I still think that.  My main argument is that if we change the semantic at all we should just change it to what we really want, which I think is to default it to service.port.  v1beta3 makes this change, but v1beta1 and 2 retain the old semantic.

**Pull request merged**